### DISCUSSION TITLE
【Question】Ragdoll not correct work on Dedicated Server.

### DISCUSSION THREAD

**Discussion by watsonsong** [Sixze/ALS-Refactored#374] [Q&A]
I found the problem is [here](https://github.com/Sixze/ALS-Refactored/blob/main/Source/ALS/Private/AlsCharacter_Actions.cpp#L815).
`
GetMesh()->AddForce((RagdollTargetLocation - GetMesh()->GetSocketLocation(PullForceSocketName)) * RagdollingState.PullForce,
		                    PullForceSocketName, true);
` 
If I comment this `AddForce` out, the ragdoll seems correctly when 'Play as Client' in a dedicated server mode. I don't know what the purpose about this AddForce.

**Comment by Sixze** [comment]
Ragdolls are not replicated and are poorly predictable, so this code is used to somehow synchronize their location on the server and client.

**Reply by tornellihenrique** [reply]
I think the problem might be related to non locally controlled characters.
If you have a NPC and activate its ragdoll playing as a client, it's `RagdollTargetLocation` will never be updated, so it'll stick to the first set location (the Pelvis). Which doesn't happen to SimulatedProxy local players.
And this will make the AddForce keep the Character floating oddly in the air.
You can see in the following screenshot the NPC Character and it's `RagdollTargetLocation`.

<img width="552" alt="image" src="https://github.com/Sixze/ALS-Refactored/assets/47005931/2be59f80-dbac-460d-901a-a79f1bf20c55">

I didn't tested a lot, but I might have reached to a good fix:
```cpp
// AlsCharacter_Actions.cpp
void AAlsCharacter::RefreshRagdollingActorTransform(const float DeltaTime)
{
	const auto bLocallyControlled{IsLocallyControlled() || IsPlayerControlled() || IsBotControlled()};
	const auto PelvisTransform{GetMesh()->GetSocketTransform(UAlsConstants::PelvisBoneName())};

	if (bLocallyControlled || (!bLocallyControlled && HasAuthority()))
	{
		SetRagdollTargetLocation(PelvisTransform.GetLocation());
	}

        // ...
}
```
This means basically to update the `RagdollTargetLocation` if the Character is either locally player controlled or AI controlled or let the Server handle it if it doesn't have a Controller or a PlayerState.
It's working according to my fast tests.

I'm not sure if this ok or secure to add to the main code so I'll let you decide it.

**Reply by watsonsong** [reply]
Thanks! This is what the problem I met. But the problem seems not resolve when I test the code.
I think `IsLocalController` is already true for the NPCs, which remote is SimulatedProxy, and local is Authority on dedicated server:
```
bool AController::IsLocalController() const
{
        // ...
	if (GetRemoteRole() != ROLE_AutonomousProxy && GetLocalRole() == ROLE_Authority)
	{
		// Local authority in control.
		return true;
	}

	return false;
}
```

**Reply by tornellihenrique** [reply]
Unfortunatelly, for my case, the `IsLocallyControlled` was never being true for the PlacedInWord NPCs, not even for the Authority Local Role Pawns. So this was the reason for the problem to be occuring.
Perhaps you case is slightly different from mine.
For now that update fixed my case, if I came up with a better and more detailed solution I come back and let you know.
Feel free to contact me if you wish to discuss anything.
Discord: tornelli (Tornelli#4688)

**Comment by Sixze** [comment]
Should be fixed in 6ff67887547efc61ad9e948211c407188a8a187c.

**Reply by watsonsong** [reply]
I upgrade and apply to this fixed, but the enemy on dedicated server ragdoll is still hang by the pull force.
It seems not work on the NPC and enemy whole run on server?

**Reply by Sixze** [reply]
I can no longer reproduce this issue. I tested this on bots with and without an AI controller, RagdollTargetLocation is updated and replicated to players as expected.

In order for RagdollTargetLocation to be updated correctly on the server, the mesh's VisibilityBasedAnimTickOption must be set to EVisibilityBasedAnimTickOption::AlwaysTickPoseAndRefreshBones, check that this is the case for you, because judging by the screenshot, the problem is here

**Reply by watsonsong** [reply]
I found maybe an error in the new commit: `bShouldSendTargetLocation = GetLocalRole() == ROLE_AutonomousProxy || (GetLocalRole() >= ROLE_Authority && GetRemoteRole() != ROLE_AutonomousProxy)`
When I was a listen server, my pawn local role is ROLE_Authority, and remote role is ROLE_AutonomousProxy. In this case, bShouldSendTargetLocation  will be false. So there is no way to update pelvis location into RagdollTargetLocation?

When I launch as a listen server, the character always be drag to the scene zero position.

**Reply by Sixze** [reply]
I hope this issue is finally fixed, see 9b33fa98108f023dea0590a95a97aef0ba8470fb.
